File: flutter_custom_tabs_android/android/src/test/java/com/github/droibit/flutter/plugins/customtabs/core/PartialCustomTabsLauncherTest.java
package com.github.droibit.flutter.plugins.customtabs.core;

import static androidx.test.ext.truth.content.IntentSubject.assertThat;
import static com.google.common.truth.Truth.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.ArgumentMatchers.same;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

import android.app.Activity;
import android.net.Uri;

import androidx.browser.customtabs.CustomTabsIntent;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;
import org.robolectric.annotation.Config;

@RunWith(AndroidJUnit4.class)
@Config(manifest = Config.NONE)
public class PartialCustomTabsLauncherTest {
    @Rule
    public MockitoRule mockitoRule = MockitoJUnit.rule();

    @Mock
    private Activity activity;

    @InjectMocks
    private PartialCustomTabsLauncher launcher;

    @Test
    public void launch_withValidCustomTabsIntent_returnsTrue() {
        final CustomTabsIntent customTabsIntent = new CustomTabsIntent.Builder()
                .setInitialActivityHeightPx(100)
                .build();

        final Uri uri = Uri.parse("https://example.com");
        final boolean result = launcher.launch(activity, uri, customTabsIntent);
        assertThat(result).isTrue();
        assertThat(customTabsIntent.intent).hasData(uri);
        verify(activity).startActivityForResult(same(customTabsIntent.intent), eq(1001));
    }

    @Test
    public void launch_withoutRequiredExtras_returnsFalse() {
        final CustomTabsIntent customTabsIntent = new CustomTabsIntent.Builder()
                .build();

        final Uri uri = Uri.parse("https://example.com");
        final boolean result = launcher.launch(activity, uri, customTabsIntent);
        assertThat(result).isFalse();

        verify(activity, never()).startActivityForResult(any(), anyInt());
    }
}

File: flutter_custom_tabs_android/android/src/test/kotlin/com/github/droibit/flutter/plugins/customtabs/core/PartialCustomTabsLauncherTest.kt
package com.github.droibit.flutter.plugins.customtabs.core

import android.app.Activity
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.net.toUri
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.ext.truth.content.IntentSubject.assertThat
import com.google.common.truth.Truth.assertThat
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit4.MockKRule
import io.mockk.verify
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.annotation.Config

@RunWith(AndroidJUnit4::class)
@Config(manifest = Config.NONE)
class PartialCustomTabsLauncherTest {
    @get:Rule
    val mockkRule = MockKRule(this)

    @MockK(relaxed = true)
    private lateinit var activity: Activity

    @InjectMockKs
    private lateinit var launcher: PartialCustomTabsLauncher

    @Test
    fun launch_withValidCustomTabsIntent_returnsTrue() {
        val customTabsIntent = CustomTabsIntent.Builder()
            .setInitialActivityHeightPx(100)
            .build()

        val uri = "https://example.com".toUri()
        val result = launcher.launch(activity, uri, customTabsIntent)
        assertThat(result).isTrue()
        assertThat(customTabsIntent.intent).hasData(uri)

        verify { activity.startActivityForResult(refEq(customTabsIntent.intent), eq(1001)) }
    }

    @Test
    fun launch_withoutRequiredExtras_returnsFalse() {
        val customTabsIntent = CustomTabsIntent.Builder().build()

        val uri = "https://example.com".toUri()
        val result = launcher.launch(activity, uri, customTabsIntent)
        assertThat(result).isFalse()

        verify(exactly = 0) { activity.startActivityForResult(any(), any()) }
    }
}